### PR TITLE
HWKAPM-625 Retrieve business transactions dynamically, based on recor…

### PR DESCRIPTION
**Pull request by objectiser**
…ded trace fragments, rather than just the list of defined configurations

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay Could you please review?

Reviewer notes:
* previously the filter sidebar only listed business transactions that had been configured in the server
* this PR instead uses the list of business transactions associated with fragments recorded that match the other criteria fields
* hwkapm 593 has example fragments (json) and a script to load them, which can be used for testing - just need to set the time frame to a month to see the information
* from a clean start, this data has two business transactions 'add book' and 'get books' - which will appear in the filter after these test fragments are reported

File: tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/AnalyticsServiceRESTTest.java [hunk 1]
```diff
@@ -179,6 +179,46 @@ public void testGetBoundEndpoints() {
...
+        trace2.setBusinessTransaction("trace2");
+        trace2.setStartTime(System.currentTimeMillis() - 2000); // Within last hour
+
+        Component c2 = new Component();
+        c2.getProperties().add(new Property("prop2", "value2"));
+        trace2.getNodes().add(c2);
+
+        try {
+            publisher.publish(null, Arrays.asList(trace1, trace2));
+        } catch (Exception e1) {
```
Comment (pavolloffay) [inline tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/AnalyticsServiceRESTTest.java]: This should be a test error.

File: tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/AnalyticsServiceRESTTest.java [hunk 2]
```diff
@@ -179,6 +179,46 @@ public void testGetBoundEndpoints() {
...
+        Wait.until(() -> service.searchFragments(null, new Criteria()).size() == 2);
+
+        // Wait to result derived
+        Wait.until(() -> analytics.getTraceCompletionTimes(null, new Criteria()).size() == 2);
+
+        List<TransactionInfo> tis = analytics.getTransactionInfo(null, new Criteria());
+
+        assertNotNull(tis);
+        assertEquals(2, tis.size());
+        assertTrue(tis.get(0).getName().equals("trace1"));
```
Comment (pavolloffay) [inline tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/AnalyticsServiceRESTTest.java]: Minor thing, In this case it's maybe better to use `Assert.assertEquals(expected, actual)` .

**Comment by pavolloffay** [comment]
This is not related to this PR: I don't like that when you select some instance of business transaction others disappear from the menu. If you want to select some other business transaction you have to click to `All`.

**Comment by objectiser** [comment]
@pavolloffay Actually this can be fixed as part of this PR - the operation to request the list of business txns should really ignore the businessTransaction property in the filter criteria as it makes no sense to specify it - either that, or the UI can be updated to copy the current criteria and blank out that field? My preference is for (1).

File: api/src/main/java/org/hawkular/apm/api/services/Criteria.java [hunk 3]
```diff
@@ -58,6 +58,33 @@
...
+        this.endTime = criteria.endTime;
+        this.businessTransaction = criteria.businessTransaction;
+        this.hostName = criteria.hostName;
+        this.upperBound = criteria.upperBound;
+        this.lowerBound = criteria.lowerBound;
+        this.principal = criteria.principal;
+        this.uri = criteria.uri;
+        this.operation = criteria.operation;
+
+        criteria.properties.forEach(pc -> this.properties.add(new PropertyCriteria(pc)));
```
Comment (pavolloffay) [inline api/src/main/java/org/hawkular/apm/api/services/Criteria.java]: Isn't `this.properties = new HasSet<>(criteria.properties)` sufficient?

Comment (pavolloffay) [inline api/src/main/java/org/hawkular/apm/api/services/Criteria.java]: And the same for following collections

File: api/src/main/java/org/hawkular/apm/api/model/analytics/TransactionInfo.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+/**
+ * This class represents information related to a business transaction associated with
+ * trace fragments.
+ *
+ * @author gbrown
+ */
+public class TransactionInfo {
```
Comment (pavolloffay) [inline api/src/main/java/org/hawkular/apm/api/model/analytics/TransactionInfo.java]: Isn't name `BusinessTransactionInfo` better?

**Comment by pavolloffay** [comment]
I think it does not return correct business transactions: see provided vide: 
[apm-625.zip](https://github.com/hawkular/hawkular-apm/files/479776/apm-625.zip)

File: api/src/main/java/org/hawkular/apm/api/model/analytics/TransactionInfo.java (hunk 4 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/model/analytics/TransactionInfo.java]: Possibly, although sometimes this concept is also known as "User defined transactions", so just trying to keep generic. But may need to consider also changing in other places.

File: tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/AnalyticsServiceRESTTest.java (hunk 1 above)
Comment (objectiser) [inline tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/AnalyticsServiceRESTTest.java]: done

File: api/src/main/java/org/hawkular/apm/api/services/Criteria.java (hunk 3 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/services/Criteria.java]: no because that would just share the same contents rather than have a fresh copy.

**Comment by objectiser** [comment]
The /search URI is not associated with either of the business txns, so only shows up when selecting 'All' business transactions.

**Comment by objectiser** [comment]
@pavolloffay Thanks!

**Pull request merged**